# How to Deal with Missing Value in the Data set

In [45]:
#lets import important libraries which we are going to use
import numpy as np
import pandas as pd

In [46]:
#importing the data set
data=pd.read_csv('melbourne.csv')
data.shape

(23547, 21)

In [47]:
#lets check first 5 row of our dataset
data.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

         Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  03-09-2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  03-12-2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  04-02-2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  04-02-2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  04-03-2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.8014    144.9958  Northern Metropolitan   
1        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
2     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
3        NaN        Yarra  -37.8114    145.0116  Northern Metropolitan   
4     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [48]:
#check info of data
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23547 entries, 0 to 23546
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Suburb         23547 non-null  object 
 1   Address        23547 non-null  object 
 2   Rooms          23547 non-null  int64  
 3   Type           23547 non-null  object 
 4   Price          18396 non-null  float64
 5   Method         23547 non-null  object 
 6   SellerG        23547 non-null  object 
 7   Date           23547 non-null  object 
 8   Distance       23546 non-null  float64
 9   Postcode       23546 non-null  float64
 10  Bedroom2       19066 non-null  float64
 11  Bathroom       19063 non-null  float64
 12  Car            18921 non-null  float64
 13  Landsize       17410 non-null  float64
 14  BuildingArea   10018 non-null  float64
 15  YearBuilt      11540 non-null  float64
 16  CouncilArea    15656 non-null  object 
 17  Lattitude      19243 non-null  float64
 18  Longti

Lets check the presence of missing value in our dataset
>isnull(): Indicates the presence of Missing values in a dataset, return boolean values that means It will return True if the data is missing and False if the data is not missing.

In [49]:
# lets first check if there is any null value present in the data
data.isnull().sum().sum()

66918

>>there is 66918 no of data points are missing that is huge

In [50]:
#lets check the missing values column wise
data.isnull().sum()

Suburb               0
Address              0
Rooms                0
Type                 0
Price             5151
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          4481
Bathroom          4484
Car               4626
Landsize          6137
BuildingArea     13529
YearBuilt        12007
CouncilArea       7891
Lattitude         4304
Longtitude        4304
Regionname           1
Propertycount        1
dtype: int64

>Count:The count method returns the number of non-missing values for each column or row. By default, it operates column-wise.

In [51]:
data.count().tail()

CouncilArea      15656
Lattitude        19243
Longtitude       19243
Regionname       23546
Propertycount    23546
dtype: int64

In [52]:
#lets also check the percentage of missing values present in the data
percentage_of_missing_values = round(100*(data.isnull().sum()/len(data.index)), 2)
print(percentage_of_missing_values)

Suburb            0.00
Address           0.00
Rooms             0.00
Type              0.00
Price            21.88
Method            0.00
SellerG           0.00
Date              0.00
Distance          0.00
Postcode          0.00
Bedroom2         19.03
Bathroom         19.04
Car              19.65
Landsize         26.06
BuildingArea     57.46
YearBuilt        50.99
CouncilArea      33.51
Lattitude        18.28
Longtitude       18.28
Regionname        0.00
Propertycount     0.00
dtype: float64


We can drop/delete the missinng values only when the missing percentage is more than 50% or it is not so important
>remove the three column having 33% to 56%
BuildingArea
YearBuilt
CouncilArea

### Methods to impute missing values
dropna(): This Function is used to drops the missing values from a dataset and returns the rest of the dataframe.

fillna(): This Function is used to replace the missing value with a specified value.

In [53]:
data=data.drop('BuildingArea',axis=1)
data = data.drop('YearBuilt', axis=1)
data = data.drop('CouncilArea', axis=1)
data.columns

Index(['Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method', 'SellerG',
       'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom', 'Car',
       'Landsize', 'Lattitude', 'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

### column is deleted 
lets check rows having greater than 5 missing values in the dataset

In [54]:
len(data[data.isnull().sum(axis=1) > 5].index)

4278

In [55]:
# retaining the rows having <= 5 NaNs
data = data[data.isnull().sum(axis=1) <= 5]

In [56]:
#count the number of missing values in each rows 
data.apply(lambda x: x.count(), axis=1)

0        17
1        18
2        18
3        17
4        18
         ..
23542    17
23543    13
23544    17
23545    18
23546    17
Length: 19269, dtype: int64

In [57]:
print(100*(data.isnull().sum()/len(data.index)), 2)

Suburb            0.000000
Address           0.000000
Rooms             0.000000
Type              0.000000
Price            21.708444
Method            0.000000
SellerG           0.000000
Date              0.000000
Distance          0.000000
Postcode          0.000000
Bedroom2          1.053506
Bathroom          1.069075
Car               1.806010
Landsize          9.647621
Lattitude         0.134932
Longtitude        0.134932
Regionname        0.000000
Propertycount     0.000000
dtype: float64 2


In [58]:
## removing all the rows where there is a missing value in the Price Column
data = data[~np.isnan(data['Price'])]

In [59]:
#lets see after deletion
print(100*(data.isnull().sum()/len(data.index)), 2)
#or we can display using round as round(100*(data.isnull().sum()/len(data.index)), 2)

Suburb           0.000000
Address          0.000000
Rooms            0.000000
Type             0.000000
Price            0.000000
Method           0.000000
SellerG          0.000000
Date             0.000000
Distance         0.000000
Postcode         0.000000
Bedroom2         1.053957
Bathroom         1.067215
Car              1.763224
Landsize         9.830306
Lattitude        0.145831
Longtitude       0.145831
Regionname       0.000000
Propertycount    0.000000
dtype: float64 2


### now lets move towards Landsize which has 9% missing value

In [60]:
#lets check the description of values in the landsize column using describe() method
data['Landsize'].describe()

count     13603.000000
mean        558.116371
std        3987.326586
min           0.000000
25%         176.500000
50%         440.000000
75%         651.000000
max      433014.000000
Name: Landsize, dtype: float64

from the above data we observe that min value is 0 and max value is 433014
so we can't use mean and mode function because of huge difference
-its mosty appropriate that we remove that rows

In [61]:
data=data[~np.isnan(data['Landsize'])]
round(100*(data.isnull().sum()/len(data.index)), 2)

Suburb           0.00
Address          0.00
Rooms            0.00
Type             0.00
Price            0.00
Method           0.00
SellerG          0.00
Date             0.00
Distance         0.00
Postcode         0.00
Bedroom2         0.00
Bathroom         0.01
Car              0.46
Landsize         0.00
Lattitude        0.16
Longtitude       0.16
Regionname       0.00
Propertycount    0.00
dtype: float64

## Lets solve for low fraction misiing value

In [62]:
data[['Bathroom','Car','Lattitude','Longtitude']].describe()

Bathroom           Car     Lattitude    Longtitude
count  13602.000000  13540.000000  13581.000000  13581.000000
mean       1.534921      1.610414    -37.809204    144.995221
std        0.691834      0.962244      0.079257      0.103913
min        0.000000      0.000000    -38.182550    144.431810
25%        1.000000      1.000000    -37.856820    144.929600
50%        1.000000      2.000000    -37.802360    145.000100
75%        2.000000      2.000000    -37.756400    145.058320
max        8.000000     10.000000    -37.408530    145.526350

#  Imputing Using fillna() method
## Note
 we can impute missing Value using mean and median when
· Data is missing completely at random.

In [63]:
#lets first go for categorical columns
data['Bathroom'].fillna(data['Bathroom'].mode()[0], inplace = True)
data['Car'].fillna(data['Car'].mode()[0], inplace = True)

In [64]:
data.isnull().sum()

Suburb            0
Address           0
Rooms             0
Type              0
Price             0
Method            0
SellerG           0
Date              0
Distance          0
Postcode          0
Bedroom2          0
Bathroom          0
Car               0
Landsize          0
Lattitude        22
Longtitude       22
Regionname        0
Propertycount     0
dtype: int64

In [65]:
#now Solve for Lattitude and Longtitude

data[['Lattitude','Longtitude']].describe()

Lattitude    Longtitude
count  13581.000000  13581.000000
mean     -37.809204    144.995221
std        0.079257      0.103913
min      -38.182550    144.431810
25%      -37.856820    144.929600
50%      -37.802360    145.000100
75%      -37.756400    145.058320
max      -37.408530    145.526350

In [66]:
#as We can see there is minor difference between mean,min,25%,50%,75% and max
data['Lattitude'] = data['Lattitude'].fillna(data['Lattitude'].mean())
#or data['Longtitude'].fillna(data['Longtitude'].mean(), inplace = True)
data['Longtitude']=data['Longtitude'].fillna(data['Longtitude'].mean())

In [67]:
#check is there any missing value is avilable or not
data.isnull().sum()

Suburb           0
Address          0
Rooms            0
Type             0
Price            0
Method           0
SellerG          0
Date             0
Distance         0
Postcode         0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
Lattitude        0
Longtitude       0
Regionname       0
Propertycount    0
dtype: int64

## Here we handled all Missing Values